File: binary-search.py
def binary_search(arr, target):
    left = 0
    right = len(arr) - 1

    while left != right:
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid
        if arr[mid] < target:
            left = mid + 1
        if arr[mid] > target:
            right = mid

    return -1


sample = [3, 7, 9, 11, 12, 15, 16, 18, 20]

index = binary_search(sample, 7)

print(index)

File: lua/binary_search.lua
local input = { 1, 2, 5, 8, 9, 20, 47, 69, 82, 124 }

local binary_search = function(list, target)
	local lo = 0
	local hi = #list

	while lo < hi do
		local mid = math.floor((hi + lo) / 2)

		if list[mid] == target then
			return mid
		end

		if list[mid] < target then
			lo = mid + 1
		end

		if list[mid] > target then
			hi = mid
		end
	end

	return -1
end

local target = 825
print("sample table:", table.concat(input, ","))
print("target:", target)
local index = binary_search(input, target)
print("found target:", index)
